### PR TITLE
Add preLoginValidation hook

### PR DESCRIPTION
**Pull request by temparus**
Adds an additional hook named preLoginValidation.
This gives the possibility to update the login credentials of the user from an external source before the actual validation takes place.

### CODE REVIEW

**Comment by LukasReschke** [comment]
Can you explain a generi use-case for this?

File: core/Controller/LoginController.php [hunk 1]
```diff
@@ -216,6 +216,9 @@ public function tryLogin($user, $password, $redirect_url, $remember_login = fals
 		}
 
 		$originalUser = $user;
+
+        $this->userManager->emit('\OC\User', 'preLoginValidation', array($user, $password));
```
Comment (LukasReschke) [inline core/Controller/LoginController.php]: Indentation is wrong.

File: lib/private/Server.php [hunk 2]
```diff
@@ -295,6 +295,9 @@ public function __construct($webRoot, \OC\Config $config) {
 				/** @var $user \OC\User\User */
 				\OC_Hook::emit('OC_User', 'post_setPassword', array('run' => true, 'uid' => $user->getUID(), 'password' => $password, 'recoveryPassword' => $recoveryPassword));
 			});
+            $userSession->listen('\OC\User', 'preLoginValidation', function ($uid, $password) {
```
Comment (LukasReschke) [inline lib/private/Server.php]: Indentation is wrong.

**Review by LukasReschke — DISMISSED** [review]

**Comment by sandrolutz** [comment]
One use-case is, when the groupmemberships and login credentials are managed by another server, which is not LDAP.
In this case, it is important that the syncronization with the other server happens before any validation of the user input. Furthermore, to sync information of new users, which have no nextcloud user account yet, it is required, that the hook is being executed idependent from the entered data.

**Comment by MorrisJobke** [comment]
> We observed that the existing preLogin hook is not called in every login attempt through the web interface. This pull request fixes this.

Then add the preLogin instead of introducing a new one? Or what is the problem with that?

**Comment by sandrolutz** [comment]
That was the plan. Unfortunately, the preLogin hook doesn't always get called. If the user doesn't exist nextcloud, the hook is not called.

**Comment by sandrolutz** [comment]
Sorry @MorrisJobke. I misunderstood your last comment. I have removed the newly introduced preValidationHook and changed the places where the preLogin hook gets called so that I will be called on every login attempt.

**Comment by MorrisJobke** [comment]
@nickvergessen @rullzer @blizzz What do you think of this?

**Comment by MorrisJobke** [comment]
@temparus Could you check out and mock the emit method for the failing unit tests:

bildschirmfoto 2017-01-31 um 17 07 47

**Comment by blizzz** [comment]
Seems reasonable to me. 

@GitHubUser4234 weren't you also using this hook somewhere?

**Comment by GitHubUser4234** [comment]
> @GitHubUser4234 weren't you also using this hook somewhere?

Uhm, close, but that was a postLogin which will be triggered after a successful LDAP login to check for password expiry. Thanks though :)

**Comment by temparus** [comment]
@MorrisJobke Fixed the problem you mentioned.

IUserManager does not automatically ensure that the object extends from PublicEmitter, so I added a type check.

**Review by MorrisJobke — DISMISSED** [review]
bildschirmfoto 2017-02-01 um 12 55 54

And please sign off all your commits. See https://github.com/nextcloud/server/blob/master/CONTRIBUTING.md#sign-your-work for more details about this.

**Comment by temparus** [comment]
Does anyone know, why TESTS=caldavtester and TESTS=carddavtester are always failing?

The error report says that a bash script could not be found although the requested file is in the repository and I didn't make any changes to the dav app.

**Comment by MorrisJobke** [comment]
@LukasReschke @blizzz @schiessle Mind to check out this one?

**Review by MorrisJobke — APPROVED** [review]
Looks good and still works 👍

**Comment by MorrisJobke** [comment]
@schiessle @nickvergessen @rullzer @blizzz Please review :)

**Review by blizzz — APPROVED** [review]

**Pull request merged**